eltwise.hpp

```diff
@@ -0,0 +1,147 @@
+/******************************************************************************
+ *  Copyright (c) 2022, Xilinx, Inc.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *
+ *  1.  Redistributions of source code must retain the above copyright notice,
+ *     this list of conditions and the following disclaimer.
+ *
+ *  2.  Redistributions in binary form must reproduce the above copyright
+ *      notice, this list of conditions and the following disclaimer in the
+ *      documentation and/or other materials provided with the distribution.
+ *
+ *  3.  Neither the name of the copyright holder nor the names of its
+ *      contributors may be used to endorse or promote products derived from
+ *      this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ *  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ *  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ *  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ *  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ *  OR BUSINESS INTERRUPTION). HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ *  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+ *  OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+ *  ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ******************************************************************************/
+ 
+/******************************************************************************
+ *
+ *  Authors: Yaman Umuroglu <yamanu@amd.com>
+ *
+ *  @file eltwise.hpp
+ *
+ * Templatized streaming elementwise operator, consuming two streams and
+ * producing a third. The operator and datatypes can be customized.
+ *
+ ******************************************************************************/
+
+#ifndef ELTWISE_HPP
+#define ELTWISE_HPP
+
+ /*!
+ * \brief EltwiseFunction: General contract for elementwise functions.
+ *
+ * This class itself has no formal significance for the implementation
+ * of the elementwise function. It provides a guidence for specific elementwise function to be used in 
+ * StreamingEltwise
+ * 
+ * \tparam TIn0 Datatype of the first input to the eltwise function
+ * \tparam TIn1 Datatype of the second input to the eltwise function
+ * \tparam TOut Datatype of the output generated by the eltwise function
+ *
+ */
+template<typename TIn0, typename TIn1, typename TOut>
+class EltwiseFunction {
```

Comment:
This duplicates the `binary_function` infrastructure provided in [activations.hpp](https://github.com/Xilinx/finn-hlslib/blob/master/activations.hpp#L59).
In general, the use of lambdas for trivial function drop-ins should be preferred.